docs/releasehistory.rst

```diff
@@ -11,6 +11,34 @@ Releases follow the ``major.minor.micro`` scheme recommended by `PEP440 <https:/
 Current Development
 -------------------
 
+New features
+""""""""""""
+- `PR #433 <https://github.com/openforcefield/openforcefield/pull/433>`_: Closes
+  `Issue #25 <https://github.com/openforcefield/openforcefield/issues/25>`_ by adding
+  initial support for the
+  `LibraryCharges tag in the SMIRNOFF specification <https://open-forcefield-toolkit.readthedocs.io/en/latest/smirnoff.html#librarycharges-library-charges-for-polymeric-residues-and-special-solvent-models>`_
+  using
+  :py:class:`LibraryChargeHandler <openforcefield.typing.engines.smirnoff.parameters.LibraryChargeHandler>`.
+  For a molecule to have charges assigned using LibraryCharges, all of its atoms must be covered by
+  at least one LibraryCharge. If an atom is covered by multiple LibraryCharges, then the last
+  one read will be applied. At this time, there is no concept of "residues" during parametrization,
+  so it is not possible to parametrize some atoms in a molecule using LibraryCharges and
+  others using another method. Further, no effort is made to ensure that the net charge
```

Comment:
> Further, no effort is made to ensure that the net charge applied using LibraryCharges is the total formal charge on the molecule.

Since we have all the formal charge information for any provided molecules (from the toolkit layers), wouldn't it be trivial to add this check?

Reply:
I'd thought about it, but concluded that it would add unnecessary complexity. At some point, maybe there will be a consensus that folks want a charge normalization scheme implemented, and we can incorporate a "don't-normalize-but-just-raise-exception-if-charges-don't-add-up" option. But for now, adding this functionality would make people immediately ask for a kwarg to disable it, and then in the future we'd have to break the API or have a growing tree of "if" statements to handle that kwarg's interaction with various normalization schemes.

Reply:
This seems highly dangerous to me since a mistyped library charge definition could lead to corrupted, non-integral charges. Wouldn't we want to be conservative first and then allow a way to relax things later if we find we need to?

Reply:
> This seems highly dangerous to me since a mistyped library charge definition could lead to corrupted, non-integral charges.

I see where you're coming from -- Our API should just "do the right thing", and that is the attitude I take toward users who show up with molecules and just "want to use the most recent FF". But here we're considering a different group of users -- those writing their own SMIRKS and `LibraryCharge` entries. I think it's likely that those users would find that our "helpful" guiderails actually get in their way. 

There are several use cases where raising an error would get in the way. For example, I suspect that someone will start experimenting with fragment-based charge assignment, with the intent to perform their own normalization on the charges in the resulting molecule. They'll be frustrated when we disallow this. 

I agree that it's possible for promiscuous SMIRKS to quietly ruin a system, so we may have to ensure that the FFs _published by Open Force Field_ don't contain potentially-promiscuous library charges. But, we already let people do tremendously dangerous things like introduce wildcard parameters into their `Bonds` section, while also concatenating sections from subsequently-loaded FFs (so, if a wildcard bond-containing file is read after Parsley, all bonds will match the wildcard). The spec allows people to create the FF equivalent of nuclear fuel -- It's great if someone else is handling it, but people need to understand that messing with parameter sources is dangerous.

What I'm getting at is that this isn't a _technical_ problem -- It would be easy to add the "formal charge" check and a resulting warning or exception+override. But users already complain that we have too many warnings. And an exception+override adds complexity on our end. If we go that route, I bet that `allow_nonintegral_charges=True` will just become another `allow_undefined_stereo=True`, which already seems to be cargo-culting around in scripts.

Reply:
+1 for enforcing that the charges add up to the formal charge with the option to disable the check. I think most people (by which I mean me) would prefer to be told explicitly when the charges don't add up, thus highlighting something which is usually very wrong in the system without having to write checks for this each time a molecule is parametrized.

Reply:
I think this is especially true when the default is to take the last charge to be read!

Reply:
Let's check with @davidlmobley here too.
Since we already provide a mechanism for experimenting with charge models---specifying the charges on a whole molecule via an API---we don't want to allow someone to try to tile fragments in a way that leads to nonintegral charges on System creation and then try to go in and manually rescale things. They can already do whatever they want by specifying a precharged Moleculez where we should let them do whatever they like because that mechanism is intended for exploration.

 However, I know for certain one of us is going to make a mistake and accidentally mistype a SMARTS by accident and end up with nonintegral charges. We need this check to ensure we are not screwing this up. So it would seem prudent to add the check by default for molecules typed via LibraryCharge.

Reply:
I talked to Jeff about this in person and actually was inclined to agree with his viewpoint for all the reasons he listed, and also worried that additional developer time he'd spend trying to keep people from making mistakes would be time taken away from scientifically higher priority tasks like BCCs and WBOs. But I don't feel that strongly about it.

Reply:
Since these library charges are going to be written by hand, we would have to implement a mechanism outside of ForceField to check for this if we don't include it within ForceField. So either way, he's going to need to write this code. I'd prefer we use it to avoid being hoist by our own petard by baking it in here...

Reply:
I appreciate the thoughtful replies on this topic. This may just be an area where my perspective as a developer is different that the perspective of the users. 

I'll go ahead and implement a check for nonintegral charge assignment, which can optionally be disabled by a keyword argument (`allow_nonintegral_charges`). To future-proof this a bit, I'll have it apply to all charge methods (so, it will run in `ElectrostaticsHandler.postprocess_system`).

Reply:
That sounds like a reasonable approach!

Reply:
Ok, I've added a check to ensure that the partial charges on all particles in a Molecule sum up to its formal charge.  This can be overridden with the kwarg `allow_nonintegral_charge=True`.